### PR TITLE
[ELF] Attempt to set the OS when using 'makeTriple()'

### PR DESCRIPTION
**Pull request by jhuber6**
Summary:
This patch fixes up the `makeTriple()` interface to emit append the
operating system information when it is readily avaialble from the ELF.
The main motivation for this is so the GPU architectures can be easily
identified correctly when given and ELF. E.g. we want
`amdgpu-amd-amdhsa` as the output and not `amdgpu--`.

This required adding support for the CUDA OS/ABI, which is easily found
to be `0x33` when using `readelf`.

### CODE REVIEW

**Comment by jhuber6** [comment]
I intentionally avoided reformatting the `llvm-readobj` code to better match the preexisting style.

**Review by arsenm — COMMENTED** [review]
Is this testable?

**Comment by jhuber6** [comment]
> Is this testable?

There's not any code that tests it right now. I think it's annoying to test because it requires a unit test and a working binary. For non-NVPTX (because of PTX) we could probably just run the backend on an empty LLVM-IR file if nothing else if we really wanted to test it. It's a lot of effort however. 

There's a use of this in `llvm-objdump` but it doesn't print it anywhere we could check it doesn't seem.

File: llvm/unittests/Object/ELFObjectFileTest.cpp [hunk 1]
```diff
@@ -287,6 +291,30 @@ TEST(ELFObjectFileTest, MachineTestForXtensa) {
...
+  for (auto [Machine, OS, Triple] : Formats) {
+    const DataForTest D(ELF::ELFCLASS64, ELF::ELFDATA2LSB, Machine, OS,
+                        ELF::EF_AMDGPU_MACH_AMDGCN_LAST);
+    Expected<std::unique_ptr<ObjectFile>> ELFObjOrErr =
+        object::ObjectFile::createELFObjectFile(
+            MemoryBufferRef(toStringRef(D.Data), "dummyELF"));
+    ASSERT_THAT_EXPECTED(ELFObjOrErr, Succeeded());
+
+    auto &ELFObj = **ELFObjOrErr;
+    ASSERT_EQ(ELFObj.makeTriple().getTriple(), Triple);
```
Comment (arsenm) [inline llvm/unittests/Object/ELFObjectFileTest.cpp:314]: EXPECT_EQ is sufficient. Also I think you need to swap the order to get the correct failure message

**Review by jlebar — APPROVED** [review]

**Comment by MaskRay** [comment]
LGTM

**Pull request merged**